File: Source/Library/iap/src/main/java/com/philips/cdp/di/iap/screens/WebFragment.java
/**
 * (C) Koninklijke Philips N.V., 2015.
 * All rights reserved.
 */

package com.philips.cdp.di.iap.screens;

import android.annotation.TargetApi;
import android.graphics.Bitmap;
import android.net.http.SslError;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.SslErrorHandler;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.philips.cdp.di.iap.R;
import com.philips.platform.uid.view.widget.ProgressBar;

public abstract class WebFragment extends InAppBaseFragment {

    public static final String TAG = WebFragment.class.getName();
    protected WebView mWebView;
    private String mUrl;
    private ProgressBar mProgress;
    private boolean mShowProgressBar = true;


    @Override
    public View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        ViewGroup viewGroup = (ViewGroup) inflater.inflate(R.layout.iap_web_payment, container, false);

        mWebView = (WebView) viewGroup.findViewById(R.id.wv_payment);
        mWebView.setWebViewClient(new IAPWebViewClient());
        mWebView.getSettings().setJavaScriptEnabled(false);

        mProgress = createCustomProgressBar(viewGroup,MEDIUM);
        mProgress.setVisibility(View.GONE);

        mUrl = getWebUrl();
        return viewGroup;
    }

    void initProgressBar(){

    }
    @Override
    public void onViewCreated(final View view, final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mWebView.loadUrl(mUrl);
    }

    @Override
    public void onResume() {
        super.onResume();
        mWebView.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        mWebView.onPause();
    }

    protected abstract String getWebUrl();

    protected boolean shouldOverrideUrlLoading(final String url) {
        return false;
    }

    private boolean shouldHandleError(final int errorCode) {
        return (errorCode == WebViewClient.ERROR_CONNECT
                || errorCode == WebViewClient.ERROR_BAD_URL
                || errorCode == WebViewClient.ERROR_TIMEOUT
                || errorCode == WebViewClient.ERROR_HOST_LOOKUP);
    }

    private class IAPWebViewClient extends WebViewClient {

        @SuppressWarnings("deprecation")
        @Override
        public boolean shouldOverrideUrlLoading(final WebView view, final String url) {
            return WebFragment.this.shouldOverrideUrlLoading(url);
        }

        @TargetApi(Build.VERSION_CODES.N)
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            return WebFragment.this.shouldOverrideUrlLoading(request.getUrl().toString());
        }

        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            mProgress.setVisibility(View.VISIBLE);
            super.onPageStarted(view, url, favicon);
        }

        @Override
        public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
            handler.proceed(); // Ignore SSL certificate errors
        }

        @SuppressWarnings("deprecation")
        @Override
        public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
        }

        @TargetApi(android.os.Build.VERSION_CODES.M)
        @Override
        public void onReceivedError(WebView view, WebResourceRequest req, WebResourceError rerr) {
            // Redirect to deprecated method, so you can use it in all SDK versions
            if (rerr != null && shouldHandleError(rerr.getErrorCode())) {
                if (isVisible()) {
                    onReceivedError(view, rerr.getErrorCode(), rerr.getDescription().toString(), req.getUrl().toString());
                }
            }
        }

        @Override
        public void onPageFinished(final WebView view, final String url) {
            if (mProgress != null && mShowProgressBar) {
                mShowProgressBar = false;
                mProgress.setVisibility(View.GONE);
            }
            super.onPageFinished(view, url);
        }
    }
}

File: Source/mec/Source/Library/mec/src/main/java/com/philips/cdp/di/mec/screens/catalog/WebFragment.kt
package com.philips.cdp.di.mec.screens.catalog

import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.net.http.SslError
import android.os.Build
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.*
import android.widget.RelativeLayout

import com.philips.cdp.di.mec.R
import com.philips.cdp.di.mec.screens.MecBaseFragment


abstract class WebFragment : MecBaseFragment() {

    val TAG = WebFragment::class.java.name
     lateinit var mWebView: WebView
    private var mUrl: String? = null
    private var mParentContainer: RelativeLayout? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val viewGroup = inflater.inflate(R.layout.mec_web_fragment, container, false) as ViewGroup
        mParentContainer = viewGroup.findViewById<View>(R.id.mec_web_container) as RelativeLayout
        mWebView = viewGroup.findViewById<View>(R.id.mec_webView) as WebView
        mWebView.webViewClient = IAPWebViewClient()
        mWebView.settings.javaScriptEnabled = isJavaScriptEnable()
        mWebView.settings.saveFormData = false
        createCustomProgressBar(mParentContainer, MEDIUM)

        mUrl = getWebUrl()
        return viewGroup
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mWebView.loadUrl(mUrl)
    }

    override fun onResume() {
        super.onResume()
        //setCartIconVisibility(false)
        mWebView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mWebView.onPause()
    }

    protected abstract fun isJavaScriptEnable(): Boolean

    protected abstract fun getWebUrl(): String?

    protected fun shouldOverrideUrlLoading(url: String): Boolean {
        if (url == null) return false

        try {
            if (url.startsWith("http:") || url.startsWith("https:")) {
                return true
            } else {
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                startActivity(intent)
                return true
            }
        } catch (e: Exception) {
            // Avoid crash due to not installed app which can handle the specific url scheme
            return false
        }

    }

    private fun shouldHandleError(errorCode: Int): Boolean {
        return (errorCode == WebViewClient.ERROR_CONNECT
                || errorCode == WebViewClient.ERROR_BAD_URL
                || errorCode == WebViewClient.ERROR_TIMEOUT
                || errorCode == WebViewClient.ERROR_HOST_LOOKUP)
    }

    private inner class IAPWebViewClient : WebViewClient() {

        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            return this@WebFragment.shouldOverrideUrlLoading(url)
        }

        @TargetApi(Build.VERSION_CODES.N)
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            return this@WebFragment.shouldOverrideUrlLoading(request.url.toString())
        }

//        override fun onPageStarted(view: WebView, url: String, favicon: Bitmap) {
//            super.onPageStarted(view, url, favicon)
//        }

        override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError) {
            super.onReceivedSslError(view, handler, error)
            handler.proceed() // Ignore SSL certificate errors
        }

        override fun onReceivedError(view: WebView, errorCode: Int, description: String, failingUrl: String) {
            hideProgressBar()
        }

        @TargetApi(Build.VERSION_CODES.M)
        override fun onReceivedError(view: WebView, req: WebResourceRequest, rerr: WebResourceError?) {
            // Redirect to deprecated method, so you can use it in all SDK versions
            if (rerr != null && shouldHandleError(rerr.errorCode)) {
                if (isVisible()) {
                    onReceivedError(view, rerr.errorCode, rerr.description.toString(), req.url.toString())
                }
            }
        }

        override fun onPageFinished(view: WebView, url: String) {
            hideProgressBar()
            super.onPageFinished(view, url)
        }
    }

    override fun handleBackEvent(): Boolean {
        if (mWebView.canGoBack()) {
            mWebView.goBack()
            return true
        }
        return false
    }
}
